module_utils/foreman_helper.py

```diff
@@ -93,6 +266,25 @@ def __init__(self, argument_spec, **kwargs):
         super(KatelloEntityAnsibleModule, self).__init__(argument_spec=args, **kwargs)
 
 
+class ForemanEntityApypieAnsibleModule(ForemanApypieAnsibleModule):
+
+    def __init__(self, argument_spec, **kwargs):
+        args = dict(
+            state=dict(choices=['present', 'absent'], default='present'),
+        )
+        args.update(argument_spec)
+        super(ForemanEntityApypieAnsibleModule, self).__init__(argument_spec=args, **kwargs)
+
+    def parse_params(self):
+        module_params = super(ForemanEntityApypieAnsibleModule, self).parse_params()
+        state = module_params.pop('state')
+        return (module_params, state)
+
+
+def sanitize_entity_dict(entity_dict, name_map):
+    return {value: entity_dict[key] for key, value in name_map.items() if key in entity_dict}
```

Comment:
How about:
```suggestion
    result = {}
    for (key, (value, normalization)) in name_map.items():
        if key in entity_dict:
            if normalization:
                result[value] = module.find_resources(normalization, entity_dict[key], thin=True)
            else:
                result[value]: entity_dict[key]

    return result
```

Though it doesn't deal with the by name/id problem, it does allow dropping a lot of code from individual modules. Perhaps the mapping needs to be a bit more expressive though.

Reply:
sanitize_entity_dict is something we also use in the nailgun based modules, so I'd prefer not to touch it just yet. but I agree, there is certainly room for further improvements